correct_periscope_drift.py

```diff
@@ -0,0 +1,421 @@
+    #
+    return mypars
+
+
+# Display parameter info to the user.
+# The format and choice of information to display is up to you;
+# I've chosen to include some ancilalry information but it
+# depends on the tool what should be used.
+#
+def display_start_info(opts):
+    v1("Running: {0}".format(opts["progname"]))
+    v2("  version = {0}".format(VERSION))
+    v2("with parameters:")
+    v2("  infile={0}".format(opts["infile"]))
+    v2("  evtfile={0}".format(opts["evtfile"]))
+    v2("  outfile={0}".format(opts["outfile"]))
+    # probably other values here too
+    v2("  verbose={0}".format(opts["verbose"]))
+    v2("  and ASCDS_INSTALL is {0}".format(os.environ["ASCDS_INSTALL"]))
+    v2("-" * 60)
+
+
+def equatorial2transform(ra, dec, roll):
+    """Construct the transform/rotation matrix from RA,Dec,Roll (given in degrees)
+    :returns: transform matrix
+    :rtype: Nx3x3 numpy array
+
+    """
+    ra = np.radians(ra)
+    dec = np.radians(dec)
+    roll = np.radians(roll)
+    ca = np.cos(ra)
+    sa = np.sin(ra)
+    cd = np.cos(dec)
+    sd = np.sin(dec)
+    cr = np.cos(roll)
+    sr = np.sin(roll)
+    # This is the transpose of the transformation matrix (related to
+    # translation of original perl code
+    rmat = np.array(
+        [[ca * cd,                  sa * cd,                sd     ],
+         [-ca * sd * sr - sa * cr, -sa * sd * sr + ca * cr, cd * sr],
+         [-ca * sd * cr + sa * sr, -sa * sd * cr - ca * sr, cd * cr]])
+
+    return rmat.transpose()
+
+
+def radec2eci(ra, dec):
+    """
+    Convert from RA,Dec to ECI.  The input ``ra`` and ``dec`` values can be 1-d
+    arrays of length N in which case the output ``ECI`` will be an array with
+    shape (3,N).
+
+    Borrowed from Ska.quatutil
+
+    :param ra: Right Ascension (degrees)
+    :param dec: Declination (degrees)
+    :returns: numpy array ECI (3-vector or 3xN array)
+    """
+    r = np.radians(ra)
+    d = np.radians(dec)
+    return np.array([np.cos(r) * np.cos(d), np.sin(r) * np.cos(d), np.sin(d)])
+
+
+#def roll_range(asp_roll):
+#    max_roll = np.max(asp_roll)
+#    min_roll = np.min(asp_roll)
+#    if (max_roll - min_roll) > 180:
+#        roll_180 = asp_roll.copy()
+#        roll_180[roll_180 > 180] -= 360
+#        max_roll = np.max(roll_180)
+#        min_roll = np.min(roll_180)
+#        if max_roll - min_roll > 180:
+#            raise ValueError("> 180 deg roll range")
+#    return max_roll - min_roll
+#
+
+def extract_events(event_file, src_x, src_y, src_radius):
+    """
+    Get events from specified source circle
+
+    :param event_file: Chandra event 1 or 2 file
+    :param src_x: Sky X coordinate of source region to extract
+    :param src_y: Sky Y coordinate of source region to extract
+    :param src_radius: Source region/circle radius in pixels
+    :returns: CRATE of events
+    """
+    regstring = "circle({},{},{})".format(src_x, src_y, src_radius)
+    events = pycrates.read_file("{}[sky={}]".format(
```

Comment:
Something like this should work:



diff --git a/correct_periscope_drift.py b/correct_periscope_drift.py
index 443b6bc..bf5c194 100755
--- a/correct_periscope_drift.py
+++ b/correct_periscope_drift.py
@@ -327,8 +327,13 @@ def main(opt):
      if asol_obsid != evt_obsid:
          v1("Error Aspect solution obsid {} != event file obsid 
{}".format(asol_obsid, evt_obsid))

-    evt_ra = events.get_column('RA').values
-    evt_dec = events.get_column('Dec').values
+
+    wcs = events.get_transform("eqpos")
+    evt_x = events.get_column('x').values
+    evt_y = events.get_column('y').values
+    rd = wcs.apply(zip(evt_x,evt_y))
+    evt_ra = rd[:,0]
+    evt_dec = rd[:,1]
      evt_times = events.get_column('Time').values

      # Limit to only using events contained within the range of the aspect solution



Since 'x' and 'y' are actual columns in the file, they aren't affected by that bug.





On 06/23/16 12:03, Kenny Glotfelty wrote:
> Hi,
>
> I found the existing bug report I was thinking of
>
> https://icxc.harvard.edu/pipe/ascds_help/2013a/0315.html
>
> and unfortunately, the bug is intermittent.  I don't see it in your example, 
> but it could show up randomly.
>
> We should be able to work-around it.  We can get the X and Y values, and apply 
> the WCS transform directly.  I'll mock something up (it won't be too many 
> lines of code.)
>
>
> Kenny
>
>
>
>
> On 06/22/16 15:01, Jean Connelly wrote:
>>
>> In correct_periscope_drift.py 
>> <https://github.com/sot/correct_periscope_drift/pull/1#discussion_r68114634>:
>>
>> > +#            raise ValueError("> 180 deg roll range")
>> > +#    return max_roll - min_roll
>> > +#
>> > +
>> > +def extract_events(event_file, src_x, src_y, src_radius):
>> > +    """
>> > +    Get events from specified source circle
>> > +
>> > +    :param event_file: Chandra event 1 or 2 file
>> > +    :param src_x: Sky X coordinate of source region to extract
>> > +    :param src_y: Sky Y coordinate of source region to extract
>> > +    :param src_radius: Source region/circle radius in pixels
>> > +    :returns: CRATE of events
>> > +    """
>> > +    regstring = "circle({},{},{})".format(src_x, src_y, src_radius)
>> > +    events = pycrates.read_file("{}[sky={}]".format(
>>
>> Yes. I'll fix the radius label in the par file. Regarding the second item, is 
>> that an action for SDS? Should I open a ticket of some kind?
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/sot/correct_periscope_drift/pull/1/files/34174a7c3481614baf72b9123ac32d7cb528beb3#r68114634>, 
>> or mute the thread 
>> <https://github.com/notifications/unsubscribe/AGWoI3tJ7WYvr5xpUAlkjs6wKBB4s6sAks5qOYaQgaJpZM4IyK7D>.
>>
>